### ISSUE TITLE
LevelDB compaction performance issue with large worlds (especially converted worlds)

### ISSUE THREAD

**Issue by dktapps** [pmmp/PocketMine-MP#6580]
## Problem description
LevelDB automatic level compaction tends to continuously hammer the disk with larger worlds (several GB).

Compacting 10 GB of data (at level 4) can take around 30 minutes of continuous I/O usage. LevelDB attempts to distribute this work in order to not max out I/O, but this appears to manifest as continuous background load on disks.

## Wtf is compaction? Why is this happening?
Compaction orders all the data keys in a DB level according to the DB's comparator (by default a simple bytewise comparator that sorts like alphabetically but for byte values).
When a level in the DB fills up, the data from the lower level is merged with the higher level, and the higher level's data may be rebuilt to order it correctly.

During compaction, any files whose key ranges do not overlap with the higher level are directly moved to the higher level without rebuilding. This is cheap, so we want this to be the norm.
Files whose key ranges _do_ overlap are costly to merge into the higher level, because the new data must be sorted with the set of existing data in the higher level.

**This means that key structure is critical to getting best compaction performance.** Keys that are accessed at similar times (e.g. chunks with similar coordinates) should have leading key bits that are as similar as possible to minimize range overlap.

However, thanks to Mojang's choice of key format, the data from lower levels is practically guaranteed to overlap because the key ordering is so disorganized relative to actual data usage patterns. This means that every compaction is I/O intensive. This is a big problem with large worlds, where the I/O cost becomes very obvious and sometimes continuous.

There are two main problems:
- Keys are formatted XXXXZZZZ, so chunks with the same X coordinate will appear next to each other regardless of how far apart they are in the actual world. e.g. if you have chunks `0,0`,`0,1`,`1,0`,`1,1`, chunk `0,2` will land in the middle of this order, requiring a lot of data to be moved. Scale this up 10 million times and you can see why large DBs are a problem.
- Keys use little-endian packed int32s, meaning that chunks with similar coordinates will appear about as far apart in the DB as they possibly could (e.g. X coordinates would be ordered like 0, 256, 65536, 1, 257, 65537, etc) since the keys are compared byte-by-byte. This is like having an index at a library having the names of books sorted according to `strrev(name)`.

## Proposed solution

There's a couple of ways to solve this problem:
1) Use a custom comparator to sort the keys optimally - Not super portable because the whole DB must be rebuilt if comparator was changed
2) Change the key structure -also not ideal, would _also_ require a full DB rebuild
 2a) Use Z-order (morton) codes or another space-filling curve - this would structure keys to look like XZXZXZXZ at the bit level instead of XXXXZZZZ.
 2b) Encode in big-endian to make sure related chunks data appears nearer each other. Little-endian doesn't make sense for this case.
3) Split the DB up into "regions" similar to Anvil & MCRegion. This would avoid the key overlap problem to a large extent by ensuring that these overlapping keys don't appear in the same DB in the first place.

The bottom line is that there's no way to fix this without deviating from the Mojang world format and forcing users to rebuild their worlds.
Really, this is a change that Mojang should make, but I doubt it'll ever happen, since vanilla worlds rarely grow big enough for this kind of problem to show up.

## Alternative solutions that don't require API changes
There doesn't seem to be a good alternative solution to this problem. I considered having compactions triggered more frequently and on specific key sets, but I don't think it will address the problem of constantly hammering I/O. In addition, compactions by key ranges aren't super helpful anyway on account of poor cache locality.

**Comment by dktapps** [comment]
Increasing `max_file_size` in LevelDB might also be an option, but this currently isn't supported by `php-leveldb`. It's also not clear how much this would just be kicking the can down the road.

**Comment by dktapps** [comment]
After some thought I think I understand the issue better:

During compaction, any files whose key ranges do not overlap with the higher level are directly moved to the higher level without rebuilding. However, thanks to Mojang's choice of key format, the sort order couldn't be more suboptimal if it tried.

There are two main problems:
- Keys are formatted XXXXZZZZ, so chunks with the same X coordinate will appear next to each other regardless of how far apart they are in the actual world
- Keys use little-endian packed int32s, meaning that chunks with similar coordinates will appear about as far apart in the DB as they possibly could (e.g. X coordinates would be ordered like 0, 256, 65536, 1, 257, 65537, etc)

This poor key structure means that during compaction, the new level's key ranges are practically guaranteed to overlap with the higher level, meaning that .ldb files from the below level cannot be directly moved to the higher level and the whole level basically always needs to be completely rebuilt so that this suboptimal ordering can be maintained. This starts becoming a noticeable issue once world size exceeds 100MB and probably even earlier.

There's a couple of ways to solve this problem:
1) Use a custom comparator to sort the keys optimally - Not super portable because the whole DB must be rebuilt if comparator was changed
2) Change the key structure to use something like XZXZXZXZ (in big-endian) - essentially morton codes - also not ideal, would _also_ require a full DB rebuild

The bottom line is that there's no way to fix this without deviating from the Mojang world format and forcing users to rebuild their worlds.
Really, this is a change that Mojang should make, but I doubt it'll ever happen, since vanilla worlds rarely grow big enough for this kind of problem to show up.

**Comment by dktapps** [comment]
Sadly local experiments with improved key structure didn't make a huge difference. It definitely improved things but not nearly as much as I'd hoped.

**Comment by dktapps** [comment]
The least destructive way to do this seems to be to split up worlds into DBs with NxN chunks to get around the cache locality issue and avoid unnecessary movement of dark data, but to keep the crappy Mojang key format for tool compatibility.

I would've liked to do a key format change too, for performance reasons, but if we stick to the Mojang key structure, the individual region DBs would still be openable by Minecraft and community tools as if they were full LevelDB worlds, which would likely be useful for debugging and world building. It would remove some of the performance benefit, but I think the ability to still use pre-existing tools would probably be preferable for the community.
